src/java.base/share/classes/java/util/stream/GathererOp.java

```diff
@@ -0,0 +1,745 @@
+        public void begin(long size) {
+            final var initializer = gatherer.initializer();
+            if (initializer != Gatherer.defaultInitializer()) // Optimization
+                state = initializer.get();
+            sink.begin(size);
+        }
+
+        @Override
+        public void accept(T t) {
+            /* Benchmarks have indicated that doing an unconditional write to
+             * `proceed` is more efficient than branching.
+             * We use `&=` here to prevent flips from `false` -> `true`.
+             *
+             * As of writing this, taking `greedy` or `stateless` into
+             * consideration at this point doesn't yield any performance gains.
+             */
+            proceed &= integrator.integrate(state, t, this);
+        }
+
+        @Override
+        public boolean cancellationRequested() {
+            return cancellationRequested(proceed);
+        }
+
+        private boolean cancellationRequested(boolean knownProceed) {
+            return !(knownProceed && (!sink.cancellationRequested() || (proceed = false)));
+        }
+
+        @Override
+        public void end() {
+            final var finisher = gatherer.finisher();
+            if (finisher != Gatherer.<A, R>defaultFinisher()) // Optimization
+                finisher.accept(state, this);
+            sink.end();
+            state = null; // GC assistance
+        }
+
+        // Gatherer.Sink contract below:
+
+        @Override
+        public boolean isRejecting() {
+            return !proceed;
+        }
+
+        @Override
+        public boolean push(R r) {
+            var p = proceed;
+            if (p)
+                sink.accept(r);
+            return !cancellationRequested(p);
+        }
+    }
+
+    private static int opFlagsFor(Integrator<?, ?, ?> integrator) {
+        return integrator instanceof Integrator.Greedy<?, ?, ?>
+                ? GREEDY_FLAGS : SHORT_CIRCUIT_FLAGS;
+    }
+
+
+    private final static int DEFAULT_FLAGS =
+            StreamOpFlag.NOT_SORTED | StreamOpFlag.NOT_DISTINCT |
+                    StreamOpFlag.NOT_SIZED;
+
+    private final static int SHORT_CIRCUIT_FLAGS =
+            DEFAULT_FLAGS | StreamOpFlag.IS_SHORT_CIRCUIT;
+
+    private final static int GREEDY_FLAGS =
+            DEFAULT_FLAGS;
+
+    final Gatherer<T, A, R> gatherer;
+
+    /*
+     * This constructor is used for initial .gather() invocations
+     */
+    private GathererOp(ReferencePipeline<?, T> upstream, Gatherer<T, A, R> gatherer) {
+        /* TODO this is a prime spot for pre-super calls to make sure that
+         * we only need to call `integrator()` once.
+         */
+        super(upstream, opFlagsFor(gatherer.integrator()));
+        this.gatherer = gatherer;
+    }
+
+    /*
+     * This constructor is used when fusing subsequent .gather() invocations
+     */
+    @SuppressWarnings("unchecked")
+    private GathererOp(Gatherer<T, A, R> gatherer, GathererOp<?, ?, T> upstream) {
```

Comment:
Is it not better to use a static method named by example `fuse`, i don't understand why it has to be a constructor.

Reply:
It is only ever exposed via a static method.